### ISSUE TITLE
Compatibility with newer Thunderbird releases?

### ISSUE THREAD

**Issue by jikamens** [trlkly/dorando-keyconfig#16]
Are you planning on releasing a new release of this add-on which is compatible with newer Thunderbird releases?

**Comment by trlkly** [comment]
No plans to make it a web extension. Do plan to do minimal maintenance to
keep it compatible with XUL-compatible apps.

I don't test often on Thunderbird. Is it not working?

**Comment by jikamens** [comment]
Not sure whether it's working, but the compatible version number on AMO is lower than the current released version of Thunderbird. Also, upcoming Thunderbird releases will desupport a bunch of depreciated XUL interfaces and JavaScript constructs, so if you're not testing the add-on on nightly or early-bird builds and fixing issues that crop up, it's pretty likely that it'll break when those changes are released to everybody.

**Comment by trlkly** [comment]
Do you have a link to more info on the changes, or the plan of what will be
removed? My understanding previously was that Thunderbird was not removing
any XUL support at all, and thus any necessary maintenance support would be
minor.

Let me put everything out there. I did not develop this addon. I just did
some minimal changes to get it where it would be signed by Mozilla and
could be hosted on their site, so more people could find what is still the
best keyboard shortcut solution for Firefox.

I can understand code on a micro level, but I do not actually understand
how the code works in its entirety. I am thus limited in what I can do to
fix it. I also have ongoing health problems that make spending too much
time on this prohibitive.

I definitely want this addon to keep working as long as it can. I will test
on Thunderbird a bunch and hope I can figure things out. However, you seem
quite knowledgeable about what's going on.

If you are interested in helping, I am very interested. And if you want to
fork this, I would gladly do whatever I could to turn everything over to
you.

This addon will stop working once Firefox 52 ESR is finished. Whether it
remains usable on the next released Thunderbird depends on exactly what
happens.

**Comment by jikamens** [comment]
>Do you have a link to more info on the changes, or the plan of what will be removed?

I honestly don't know if they've documented it anywhere. Here are the ones I know of:

* `Components.interfaces.nsIPrefBranch2` is gone and merged into `Components.interfaces.nsIPrefBranch`.
* `Components.interfaces.nsILocalFile` is gone and merged into `Components.interfaces.nsIFile`.
* `Components.interfaces.nsILocaleService` is gone. There isn't anything replacing it with the same API. Instead, you have to update your code to use the different interface provided by `Components.interfaces.mozILocaleService`.
* `Components.interfaces.nsIScriptableDateFormat` is gone. There's no direct API replacement, so you have to update your code to use, e.g., the `toLocaleString` function on `Date` objects.
* The construct `for each (var item in array)` no longer works. Use `for (var item of array)` instead.
* The JavaScript interpreter is more strict about requiring variables to be declared.
* You can no longer write octal numbers by prefixing them with `0`. Instead, you have to prefix them with `0o'.

Like I said, these are just the ones I know about; there may be others.

In fact, there definitely are. See also https://groups.google.com/forum/#!topic/mozilla.dev.apps.thunderbird/JrpIa7Gwlq4 for a discussion of removed interfaces (but this doesn't cover the JavaScript interpreter changes).

>If you are interested in helping, I am very interested.

If you want to add me as a collaborator to the project (https://github.com/trlkly/dorando-keyconfig/settings/collaboration), I'll take a look as time permits and try to find some time to help out. Alternatively, if you just want to be done with it, you can transfer ownership to me (at the bottom of https://github.com/trlkly/dorando-keyconfig/settings) and I'll take over maintaining it. What's one more add-on when I'm [already maintaining eight](https://blog.kamens.us/send-later/#other-addons)? ;-)

**Comment by trlkly** [comment]
You honestly sound like you are in a better place to maintain this than I am. Still, I think the responsible thing to do would be to add you as a collaborator first, and then probably transfer ownership to you later.

I'd also want you on the addons.mozilla.org page, so new releases could be uploaded there. Do you have an account there under your email address listed here?  

I definitely want the addon to stay on AMO for as long as that is allowed, and not to break on any version of Firefox with XUL support. That's the main reason I went to all the trouble I did.

**Comment by jikamens** [comment]
My AMO account email address is jik@kamens.us. I'll take a look at the code when I get a chance and see what needs to be done to bring it up to compatibility with current TB.

**Comment by ElhemEnohpi** [comment]
Keyconfig has stopped working for me as of Thunderbird 57b, also in 58b. I can still open the Keyconfig window, but the list of key assignments is blank. I now get the old default Thunderbird key assignments, including the horrific single-key shortcuts! Is it just me? Can something be done?

**Comment by trlkly** [comment]
I think this is probably better handled in a new bug. This was about the version of Thunderbird back in September. I'll rename and close. 

@ElhemEnohpi  Please create a new bug for this.

**Comment by ElhemEnohpi** [comment]
Actually I'm pretty sure I'm talking about the same issue. I think the changes that @jikamens mentioned have made their way to Thunderbird beta channel, in early November, with 57.

Keyconfig stopped working completely in Firefox 57, as expected. But Thunderbird 57 is another story, it should be possible to continue compatibility, one way or another. I gues it will require some alterations regarding the above API , but I don't know if that's only in the addon, or also in Thunderbird.

On the other hand, I could be completely wrong, and maybe it's just a problem with my user profile or something... so I'll open a new bug.
